File: frontend/src/api/predictionRecords.ts
import client from './client'

export type AccuracyStatus = 'unknown' | 'accurate' | 'inaccurate'

export interface PredictionRecord {
  id: number
  stock_name: string
  stock_code: string
  forward_days: number
  current_price: number | null
  direction: string
  signal: string
  recommendation: string
  expected_price: number | null
  support: number | null
  resistance: number | null
  prediction_date: string
  accuracy: AccuracyStatus
  created_at: string
}

export interface PredictionRecordCreate {
  stock_name: string
  stock_code: string
  forward_days: number
  current_price: number | null
  direction: string
  signal: string
  recommendation: string
  expected_price: number | null
  support: number | null
  resistance: number | null
  prediction_date: string
  accuracy?: AccuracyStatus
}

export interface PredictionRecordUpdate {
  accuracy: AccuracyStatus
}

export async function listPredictionRecords(limit: number = 200): Promise<PredictionRecord[]> {
  const response = await client.get<PredictionRecord[]>('/prediction-records', {
    params: { limit }
  })
  return response.data
}

export async function createPredictionRecord(payload: PredictionRecordCreate): Promise<PredictionRecord> {
  const response = await client.post<PredictionRecord>('/prediction-records', payload)
  return response.data
}

export async function updatePredictionRecord(
  recordId: number,
  payload: PredictionRecordUpdate
): Promise<PredictionRecord> {
  const response = await client.patch<PredictionRecord>(`/prediction-records/${recordId}`, payload)
  return response.data
}

export async function deletePredictionRecord(recordId: number) {
  const response = await client.delete(`/prediction-records/${recordId}`)
  return response.data
}

File: backend/app/api/v1/prediction_records.py
"""Prediction record API endpoints."""
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession
from sqlmodel import select

from app.database import get_session
from app.models.prediction_record import PredictionRecord

router = APIRouter()


class PredictionRecordCreate(BaseModel):
    stock_name: str
    stock_code: str
    forward_days: int
    current_price: Optional[float] = None
    direction: str
    signal: str
    recommendation: str
    expected_price: Optional[float] = None
    support: Optional[float] = None
    resistance: Optional[float] = None
    prediction_date: str
    accuracy: str = Field(default="unknown", pattern="^(unknown|accurate|inaccurate)$")


class PredictionRecordUpdate(BaseModel):
    accuracy: str = Field(..., pattern="^(unknown|accurate|inaccurate)$")


@router.get("/", response_model=List[PredictionRecord])
async def list_prediction_records(
    limit: int = Query(default=200, ge=1, le=1000),
    session: AsyncSession = Depends(get_session),
):
    result = await session.execute(
        select(PredictionRecord)
        .order_by(PredictionRecord.created_at.desc())
        .limit(limit)
    )
    return result.scalars().all()


@router.post("/", response_model=PredictionRecord)
async def create_prediction_record(
    payload: PredictionRecordCreate,
    session: AsyncSession = Depends(get_session),
):
    record = PredictionRecord(**payload.model_dump())
    session.add(record)
    await session.commit()
    await session.refresh(record)
    return record


@router.patch("/{record_id}", response_model=PredictionRecord)
async def update_prediction_record(
    record_id: int,
    payload: PredictionRecordUpdate,
    session: AsyncSession = Depends(get_session),
):
    record = await session.get(PredictionRecord, record_id)
    if not record:
        raise HTTPException(status_code=404, detail="Prediction record not found")

    record.accuracy = payload.accuracy
    await session.commit()
    await session.refresh(record)
    return record


@router.delete("/{record_id}")
async def delete_prediction_record(
    record_id: int,
    session: AsyncSession = Depends(get_session),
):
    record = await session.get(PredictionRecord, record_id)
    if not record:
        raise HTTPException(status_code=404, detail="Prediction record not found")

    await session.delete(record)
    await session.commit()
    return {"message": "Prediction record deleted"}
